File: C++ Homeworks/PA1-Basic tetris/tetromino.cpp
#include "tetromino.h"

// Tetromino constructor
Tetromino::Tetromino(const tetrominos &s)
{
    if(s == tetrominos::I) {//control the enum type and set shape vector's shape
         shape = { {' ',' ',' ',' '},
                   {' ',' ',' ',' '},
                   {' ',' ',' ',' '},
                   {'I','I','I','I'} };
         shapeLetter = 'I';
      }
      else if(s == tetrominos::O) {
         shape = { {' ',' ',' ',' '},
                   {' ',' ',' ',' '},
                   {'O','O',' ',' '},
                   {'O','O',' ',' '} };
         shapeLetter = 'O';
      }
      else if(s == tetrominos::T) {
         shape = { {' ',' ',' ',' '},
                   {' ',' ',' ',' '},
                   {'T','T','T',' '},
                   {' ','T',' ',' '} };
         shapeLetter = 'T';
      }
      else if(s == tetrominos::J) {
         shape = { {' ',' ',' ',' '},
                   {' ','J',' ',' '},
                   {' ','J',' ',' '},
                   {'J','J',' ',' '} };
         shapeLetter = 'J';   
      }
      else if(s == tetrominos::L) {
         shape = { {' ',' ',' ',' '},
                   {'L',' ',' ',' '},
                   {'L',' ',' ',' '},
                   {'L','L',' ',' '} };
         shapeLetter = 'L';
      }
      else if(s == tetrominos::S) {
         shape = { {' ',' ',' ',' '},
                   {' ',' ',' ',' '},
                   {' ','S','S',' '},
                   {'S','S',' ',' '} };
         shapeLetter = 'S';
      }
      else if(s == tetrominos::Z) {
         shape = { {' ',' ',' ',' '},
                   {' ',' ',' ',' '},
                   {'Z','Z',' ',' '},
                   {' ','Z','Z',' '} };
         shapeLetter = 'Z';
      }
      for(int i = 0;i < 4;++i){//Get positions' of shape
         for(int j = 0;j < 4;++j){
            if(shape[i][j] != ' ') {
               onePos.at(0) = i;
               onePos.at(1) = j;
            shapePos.push_back(onePos);
            }
         }
      }
}

void Tetromino::Print() {
    for(int i = 0;i < 4;++i){
        for(int j = 0;j < 4;++j){
        cout<<shape[i][j];
     }
     cout<<endl;
    }
}

void Tetromino::Rotate(const char &rotation){
   if(rotation == 'R'){//Rotate the shape with swap tactic
      for(int i = 0;i < 4;++i){
         for(int j = 0;j < i;++j) {
            swap(shape[i][j],shape[j][i]);
         }
      }
      for(int i = 0;i < 4;i++) {
         swap(shape[i][0],shape[i][3]);
         swap(shape[i][1],shape[i][2]);
      }
   }
   if(rotation == 'L') {
      for(int i = 0;i < 4;++i){
         for(int j = 0;j < i;++j) {
            swap(shape[i][j],shape[j][i]);
         }
      }
      for(int i = 0;i < 4;++i) {
         swap(shape[0][i],shape[3][i]);
         swap(shape[1][i],shape[2][i]);
      }
   }
   fixShape();
}

void Tetromino::fixShape(){
   while(shape[3][0] == ' ' && shape[3][1] == ' ' && shape[3][2] == ' ' && shape[3][3] == ' '){
      vector<char> temp;//put the shape to left corner after rotate
      temp.push_back(shape[3][0]);//store last row of the shape
      temp.push_back(shape[3][1]);
      temp.push_back(shape[3][2]);
      temp.push_back(shape[3][3]);
      for(int i = 3;i > 0;--i){//swap the vector's chars
         for(int j = 0;j < 4;++j){
             shape[i][j] = shape[i - 1][j];
         }
      }
        
      shape[0][0] = temp.at(0);//put last row to first row
      shape[0][1] = temp.at(1);
      shape[0][2] = temp.at(2);
      shape[0][3] = temp.at(3);
   }
   while(shape[1][0] == ' ' && shape[2][0] == ' ' && shape[3][0] == ' ' && shape[0][0] == ' '){//same process but this time fix columns
      vector<char> temp;
      temp.push_back(shape[0][0]);
      temp.push_back(shape[1][0]);
      temp.push_back(shape[2][0]);
      temp.push_back(shape[3][0]);
      for(int i = 0;i < 3;++i){
         for(int j = 0;j < 4;++j){
             shape[j][i] = shape[j][i + 1];
         }
      }
        
      shape[0][3] = temp.at(0);
      shape[1][3] = temp.at(1);
      shape[2][3] = temp.at(2);
      shape[3][3] = temp.at(3);
   }  
}

bool Tetromino::canFit(Tetromino &other,const vector<vector<int>> &pos){
      vector<vector<char>> board(4,vector<char>(8,' '));//temp board
      vector<vector<char>> tempShape;
      tempShape = other.GetShape();
      int check = 1;
      int i = 0;
      int j = 0;
      while(i < shapePos.size()){//Put the first block to temp board
         board[shapePos[i][0]][shapePos[i][1]] = shape[shapePos[i][0]][shapePos[i][1]];
         ++i;
      }
      do{//The loop until the block find empty space
         i = 0;
         check = 1;
         while(i < 4){
            if(board[pos[i][0]][pos[i][1] + j] != ' '){//If the places where we want to put the block are not empty, it checks the column again by shifting it by 1.
               check = 0;
               ++j;
               break;
            }
            ++i;
         }
      }while(check == 0);
      i = 0;
      while(i < 4){//put second block to temp board
         board[pos[i][0]][pos[i][1] + j] = tempShape[pos[i][0]][pos[i][1]];
         ++i;
      }

      for(int i = 0;i < 4;++i){
        for(int j = 0;j < 8;++j){
        cout<<board[i][j];
      }
      cout<<endl;
      }
      for(i = j - 1;i < pos[i][1] + j;++i){ //controls if they fit without holes or not. If they don't returns false
         if(board[3][i] == ' ') return false;
      }
      return true;
}

vector<vector<char>> Tetromino::GetShape(){
   return shape;
}

vector<vector<int>> Tetromino::GetShapePos(){
   return shapePos;
}

File: Tetris using Java/Tetromino.java
/**
* @author Akif Safa Angi
* @version 1.0.0
*/
/**
 *Class of tetromino shape
*/
public class Tetromino {
    private char[][] shape = new char[4][4];
    private int[][] shapePos = new int[4][2];

    public enum tetrominos {I,O,T,J,L,S,Z} //Create strong enum
    /**
     *Constructor of class
     *@param letter type of tetromino shape
     *Create new shape and store position of shape to shapePos
    */
    public Tetromino(tetrominos letter) {
        for(int i = 0;i < 4;++i) {
            for(int j = 0;j < 4;++j) {
                shape[i][j] = ' ';
            }
        }
        if(letter == tetrominos.I) {
            shape[3][0] = 'I';
            shape[3][1] = 'I';
            shape[3][2] = 'I';
            shape[3][3] = 'I';
        }
        else if(letter == tetrominos.O) {
            shape[2][0] = 'O';
            shape[2][1] = 'O';
            shape[3][0] = 'O';
            shape[3][1] = 'O';
        }
        else if(letter == tetrominos.T) {
            shape[2][0] = 'T';
            shape[2][1] = 'T';
            shape[2][2] = 'T';
            shape[3][1] = 'T';
        }
        else if(letter == tetrominos.J) {
            shape[1][1] = 'J';
            shape[2][1] = 'J';
            shape[3][0] = 'J';
            shape[3][1] = 'J';
        }
        else if(letter == tetrominos.L) {
            shape[1][0] = 'L';
            shape[2][0] = 'L';
            shape[3][0] = 'L';
            shape[3][1] = 'L';
        }
        else if(letter == tetrominos.S) {
            shape[2][1] = 'S';
            shape[2][2] = 'S';
            shape[3][0] = 'S';
            shape[3][1] = 'S';
        }
        else if(letter == tetrominos.Z) {
            shape[2][0] = 'Z';
            shape[2][1] = 'Z';
            shape[3][1] = 'Z';
            shape[3][2] = 'Z';
        }
        int temp = 0;
        for(int i = 0;i < 4;++i){//Get positions' of shape
            for(int j = 0;j < 4;++j){
                if(shape[i][j] != ' ') {
                    shapePos[temp][0] = i;
                    shapePos[temp][1] = j;
                    ++temp;
                }
            }
        }
    }
    /**
     *Rotate the shape with swap tactic
     *@param rotation direction
    */
    public void Rotate() {
        for(int i=0;i<4;i++) {
            for(int j=i;j<4;j++) {
                char temp = shape[i][j];
                shape[i][j] = shape[j][i];
                shape[j][i] = temp;
                }
        }

        for(int i=0;i<4;i++) {  
            int low = 0, high = 3;
            while(low < high) {
                char temp = shape[i][low];
                shape[i][low] = shape[i][high];
                shape[i][high] = temp;
                ++low;
                --high;
            }
        }
        fixShape();
    }
    /**
     *Set shape to left corner
    */
    private void fixShape() {
        while(shape[3][0] == ' ' && shape[3][1] == ' ' && shape[3][2] == ' ' && shape[3][3] == ' '){
            char[] temp = new char[4];//put the shape to left corner after rotate
            temp[0] = shape[3][0];//store last row of the shape
            temp[1] = shape[3][1];
            temp[2] = shape[3][2];
            temp[3] = shape[3][3];
            for(int i = 3;i > 0;--i){//swap the vector's chars
                for(int j = 0;j < 4;++j){
                    shape[i][j] = shape[i - 1][j];
                }
            }
                
            shape[0][0] = temp[0];//put last row to first row
            shape[0][1] = temp[1];
            shape[0][2] = temp[2];
            shape[0][3] = temp[3];
        }
        while(shape[1][0] == ' ' && shape[2][0] == ' ' && shape[3][0] == ' ' && shape[0][0] == ' '){//same process but this time fix columns
            char[] temp = new char[4];
            temp[0] = shape[0][0];//store first collumn of the shape
            temp[1] = shape[1][0];
            temp[2] = shape[2][0];
            temp[3] = shape[3][0];
            for(int i = 0;i < 3;++i){
                for(int j = 0;j < 4;++j){
                    shape[j][i] = shape[j][i + 1];
                }
            }
                
            shape[0][3] = temp[0];//put first collumn to last collumn
            shape[1][3] = temp[1];
            shape[2][3] = temp[2];
            shape[3][3] = temp[3];
        }
        int temp = 0;
        for(int i = 0;i < 4;++i){//Get positions' of shape
            for(int j = 0;j < 4;++j){
                if(shape[i][j] != ' ') {
                    shapePos[temp][0] = i;
                    shapePos[temp][1] = j;
                    ++temp;
                }
            }
        }
    }
    /**
     *Return shape
     *@return shape
    */
    public char[][] GetShape(){
        return shape;
    }
    /**
     *Return position of shape shapePos
     *@return shapePos
    */
    public int[][] GetShapePos(){
        return shapePos;
    }
}